File: cases/core/capability_safe/PY-CS-001/project/agent/router.py
"""Agent router — dispatches plan steps to the appropriate tool."""

from __future__ import annotations

import logging

from agent.planner import PlanStep
from tools.command_runner import run_command

logger = logging.getLogger(__name__)


def _error_result(command: str, reason: str) -> dict:
    return {
        "command": command,
        "allowed": False,
        "return_code": None,
        "stdout": "",
        "stderr": "",
        "errors": [reason],
    }


def execute_plan(plan: list[PlanStep]) -> list[dict]:
    """Execute each step in the coding-agent plan and collect results."""
    results: list[dict] = []

    for step in plan:
        if not isinstance(step, dict):
            logger.warning("Skipping non-dict plan step: %r", step)
            continue

        action = step.get("action")
        if not isinstance(action, str):
            results.append(_error_result("", "Plan step missing valid 'action' key"))
            continue

        if action == "run_command":
            command = step.get("command")
            if not isinstance(command, str) or not command:
                results.append(_error_result("", "run_command step missing valid 'command'"))
                continue
            result = run_command(command)
            results.append(
                {
                    "command": result.command,
                    "allowed": result.allowed,
                    "return_code": result.return_code,
                    "stdout": result.stdout[:500] if result.stdout else "",
                    "stderr": result.stderr[:500] if result.stderr else "",
                    "errors": result.errors,
                }
            )
        else:
            results.append(_error_result(
                step.get("command", ""),
                f"Unknown action: {action}",
            ))

    return results

File: cases/core/synthetic_vulnerable/TS-SV-001/pr/base/agent/router.ts
/**
 * Agent router — dispatches each plan step to the appropriate tool.
 */

import { PlanStep } from "./planner";

export interface StepResult {
  step: PlanStep;
  outcome: { status: "skipped"; reason: string };
}

export async function executePlan(plan: PlanStep[]): Promise<StepResult[]> {
  const results: StepResult[] = [];

  for (const step of plan) {
    switch (step.action) {
      case "run_command": {
        results.push({
          step,
          outcome: {
            status: "skipped",
            reason: "Command execution is not supported yet",
          },
        });
        break;
      }

      // Other action types would be handled here in a full implementation.
      default:
        results.push({
          step,
          outcome: { status: "skipped", reason: `Unhandled action: ${step.action}` },
        });
    }
  }

  return results;
}
